### PR TITLE
Use the env variable TOIL_WORKDIR  (resolves #691)

### LINKED ISSUE #691: Use env variable TOIL_WORKDIR preferentially over temp

### PR DESCRIPTION
**Pull request by arkal**
Hierarchy for selection of a workdir is

config.workDir >> $TOIL_WORKDIR >> $TMPDIR

This will help with lazy setup of the toil-disk flag in mesos in future versions.

Resolves #691

### CODE REVIEW

File: src/toil/worker.py [hunk 1]
```diff
@@ -133,7 +133,10 @@ def badWorker():
 
     setLogLevel(config.logLevel)
 
-    tempRootDir = config.workDir
+    # Hierarchy for workdir is config.workDir >> $TOIL_WORKDIR >> $TMPDIR
+    # If config.workDir is None, try setting it to $TOIL_WORKDIR. If that is not set, os.getenv
+    # returns None.
+    tempRootDir = config.workDir if config.workDir else os.getenv('TOIL_WORKDIR')
```
Comment (hannes-ucsc) [inline src/toil/worker.py]: This can be shortened to `config.workDir or os.getenv('TOIL_WORKDIR')`

**Comment by hannes-ucsc** [comment]
Jenkins, test this please.

**Comment by hannes-ucsc** [comment]
Please rebase and squash.

**Comment by arkal** [comment]
@hannes-ucsc If the user hasn't provided workDir and  TOIL_WORKDIR is not set, then the fallback is TMPDIR. We were talking about eliminating fallbacks yesterday. How do you want to proceed? I think fallbacks are sort-of necessary here.

**Comment by arkal** [comment]
@hannes-ucsc I have this in 2 separate commits so we can roll back to the earlier one if you prefer. I can rebase to a single commit if this is ok.

**Comment by hannes-ucsc** [comment]
Jenkins, test this please.

File: src/toil/worker.py [hunk 2]
```diff
@@ -133,9 +134,7 @@ def badWorker():
 
     setLogLevel(config.logLevel)
 
-    tempRootDir = config.workDir
-    if tempRootDir is not None and not os.path.exists(tempRootDir):
-        raise RuntimeError("The temporary directory specified by workDir: %s does not exist" % tempRootDir)
+    tempRootDir = getToilWorkDir(config.workDir)
```
Comment (hannes-ucsc) [inline src/toil/worker.py]: Pass entire config. Extracting a specific attribute is an implementation detail of `getToilWorkDir()`

File: src/toil/common.py [hunk 3]
```diff
@@ -524,3 +525,33 @@ def parseSetEnv(l):
...
+    This is particularly useful if the user hasn't provided a value for --workDir (passed to this
+    function as configWorkDir) and the TOIL_WORKDIR environment variable is unset.  In this case,
+    the value returned points to a directory 'toil' located somewhere on the system in a location
+    reserved for temporary files (obtained from the tempfile module).  The variable is created if
+    required.
+    Hierarchy for workdir is config.workDir >> $TOIL_WORKDIR >> $TMPDIR
+    :param configWorkDir: Value passed to the program using the workDir flag
+    :return: Path to the work dir
+    """
+    configWorkDir = configWorkDir or os.getenv('TOIL_WORKDIR')
```
Comment (hannes-ucsc) [inline src/toil/common.py]: rename to workDir

File: src/toil/common.py [hunk 4]
```diff
@@ -524,3 +525,33 @@ def parseSetEnv(l):
...
+    if configWorkDir is not None and not os.path.exists(configWorkDir):
```
Comment (hannes-ucsc) [inline src/toil/common.py]: Move to end of function, remove first term in condition.

File: src/toil/common.py [hunk 5]
```diff
@@ -524,3 +525,33 @@ def parseSetEnv(l):
...
+        raise RuntimeError("The temporary directory specified by workDir: %s does not "
+                           "exist" % tempRootDir)
+    elif configWorkDir is None:
+        #  Neither --workDir not TOIL_WORKDIR is set
+        configWorkDir = os.path.join(tempfile.gettempdir(), 'toil')
```
Comment (hannes-ucsc) [inline src/toil/common.py]: This would give every run of every user script the same working directory. Clearly not what we want, or?

**Comment by hannes-ucsc** [comment]
Don't keep obsolete commits in a PR. If you want to hold on to a likely-to-be obsolete commit for backup purposes, just have a local branch point to it.

File: src/toil/common.py (hunk 5 above)
Comment (arkal) [inline src/toil/common.py]: The way it works now, worker temp directories will just get dumped into the directory pointed to by $TMPDIR (or other temp identifier). This way, it gets dumped instead into a directory called toil that may, or may not exist. We want to be able to do a folder walk on the toil workdir and if it is exactly $TMPDIR, we are going to have a lot of files messing with our estimations. This reduces that aspect to some extent.

What do you think of 'toil' + workFlowID ?

File: src/toil/worker.py (hunk 2 above)
Comment (arkal) [inline src/toil/worker.py]: This function will be used by multiple things. One of them will be executor.py that knows only config.workDIR, but doesn't know the rest of config.

File: src/toil/common.py (hunk 5 above)
Comment (hannes-ucsc) [inline src/toil/common.py]: That might work.

Then the question is if the the workflowID should also be appended to an explicitly configured --workDir.

Comment (arkal) [inline src/toil/common.py]: yes, it's worthy of discussion. A user *can* really specify a non-empty directory as the workdir.

**Comment by hannes-ucsc** [comment]
Discussed with @arkal. Decided to implement TOIL_WORKDIR and --workDir and TMPDIR all the same in that they point to a parent dir (the work dir) that contain per-workflow subdirs (the workflow dirs).

**Comment by arkal** [comment]
Jenkins, test this please.

File: src/toil/common.py [hunk 6]
```diff
@@ -524,3 +530,36 @@ def parseSetEnv(l):
...
+    the value returned points to a directory 'toil' located somewhere on the system in a location
+    reserved for temporary files (obtained from the tempfile module).  The variable is created if
+    required.
+    Hierarchy for workdir is config.workDir >> $TOIL_WORKDIR >> $TMPDIR
+    :param workflowID: Unique identifier for the workflow
+    :param workDir: Value passed to the program using the workDir flag
+    :return: Path to the work dir
+    """
+    workDir = workDir or os.getenv('TOIL_WORKDIR')
+    if workDir is None:
```
Comment (hannes-ucsc) [inline src/toil/common.py]: Looks like this doesn't do what we agreed on: https://github.com/BD2KGenomics/toil/pull/692#issuecomment-193533704

We want the workflow dir to always be created, right?

Comment (arkal) [inline src/toil/common.py]: whoops. Fixed that

**Comment by arkal** [comment]
@hannes-ucsc this is ready for review

File: src/toil/common.py [hunk 7]
```diff
@@ -524,3 +530,36 @@ def parseSetEnv(l):
...
+    (obtained from the tempfile module).  The workflow directory is created one level below workDir
+    and is named toil-<workFlowID>.
+    Hierarchy for workdir is config.workDir >> $TOIL_WORKDIR >> $TMPDIR
+    :param workflowID: Unique identifier for the workflow
+    :param configWorkDir: Value passed to the program using the --workDir flag
+    :return: Path to the workflow directory
+    """
+    workDir = configWorkDir or os.getenv('TOIL_WORKDIR') or tempfile.gettempdir()
+    if not os.path.exists(workDir):
+        # This will never happen if tempfile.tempdir() was used (which hapens only if the other 2
```
Comment (hannes-ucsc) [inline src/toil/common.py]: remove this comment

File: src/toil/common.py [hunk 8]
```diff
@@ -524,3 +530,36 @@ def parseSetEnv(l):
...
+        # were not set)
+        raise RuntimeError("The directory specified by workDir/TOIL_TEMPDIR: %s " % workDir +
```
Comment (hannes-ucsc) [inline src/toil/common.py]: --workDir or TOIL_WORKDIR

File: src/toil/common.py [hunk 9]
```diff
@@ -524,3 +530,36 @@ def parseSetEnv(l):
...
+    workflowDir = os.path.join(workDir, 'toil-%s' % workflowID)
+    try:
+        # Directory creation is atomic
+        os.mkdir(workflowDir)
+    except OSError as err:
+        if err.errno != 17:
+            # The directory exists if a previous worker set it up.
+            raise
+    else:
+        logger.info('Created the workFlow directory at %s' % workflowDir)
```
Comment (hannes-ucsc) [inline src/toil/common.py]: workflow

File: src/toil/common.py [hunk 10]
```diff
@@ -524,3 +530,36 @@ def parseSetEnv(l):
             raise ValueError('Empty name')
         d[k] = v
     return d
+
+
+def getToilWorkflowDir(workflowID, configWorkDir=None):
+    """
+    Returns a path to the directory where worker directories and the cache will be located for this
+    workflow.  This is particularly useful if the user hasn't provided a value for --workDir (passed
```
Comment (hannes-ucsc) [inline src/toil/common.py]: cut everything after the first sentence. A docstring should describe what a function does, not how.

File: src/toil/common.py [hunk 11]
```diff
@@ -524,3 +530,36 @@ def parseSetEnv(l):
...
+    to this function as workDir) and the TOIL_WORKDIR environment variable is unset.  In this case,
+    the value used for workDir is somewhere on the system in a location reserved for temporary files
+    (obtained from the tempfile module).  The workflow directory is created one level below workDir
+    and is named toil-<workFlowID>.
+    Hierarchy for workdir is config.workDir >> $TOIL_WORKDIR >> $TMPDIR
+    :param workflowID: Unique identifier for the workflow
```
Comment (hannes-ucsc) [inline src/toil/common.py]: Add type annotations, e.g `:param str workflowID`

**Comment by arkal** [comment]
Thanks @hannes-ucsc, I've made the changes.

**Comment by hannes-ucsc** [comment]
Jenkins, test this please.

File: src/toil/common.py [hunk 12]
```diff
@@ -524,3 +530,30 @@ def parseSetEnv(l):
             raise ValueError('Empty name')
         d[k] = v
     return d
+
+
+def getToilWorkflowDir(workflowID, configWorkDir=None):
+    """
+    Returns a path to the directory where worker directories and the cache will be located for this
+    workflow.
+    :param str workflowID: Unique identifier for the workflow
```
Comment (hannes-ucsc) [inline src/toil/common.py]: insert empty line here

File: src/toil/common.py [hunk 13]
```diff
@@ -524,3 +530,30 @@ def parseSetEnv(l):
...
+    :param str configWorkDir: Value passed to the program using the --workDir flag
+    :return: Path to the workflow directory
+    :rtype: str
+    """
+    workDir = configWorkDir or os.getenv('TOIL_WORKDIR') or tempfile.gettempdir()
+    if not os.path.exists(workDir):
+        raise RuntimeError("The directory specified by --workDir or TOIL_WORKDIR: %s " % workDir +
```
Comment (hannes-ucsc) [inline src/toil/common.py]: use either string concatenation or interpolation. Not both.

Comment (hannes-ucsc) [inline src/toil/common.py]: Also, a grammar purist would want the message to be `The directory specified by --workDir or TOIL_WORKDIR (%s) does not exist`. It's good to have delimiters around an interpolated string so as to better identify empty strings.

File: src/toil/common.py [hunk 14]
```diff
@@ -524,3 +530,30 @@ def parseSetEnv(l):
...
+                           "does not exist.")
+    # create the workflow dir
```
Comment (hannes-ucsc) [inline src/toil/common.py]: # Create

File: src/toil/jobStores/abstractJobStore.py [hunk 15]
```diff
@@ -79,6 +79,7 @@ def __init__(self, config=None):
             config.workflowID = str(uuid4())
             self.__config = config
             self.writeConfigToStore()
+        logger.info('The workflow ID is: %s' %self.__config.workflowID)
```
Comment (hannes-ucsc) [inline src/toil/jobStores/abstractJobStore.py]: Should this be DEBUG level? And again `"The workflow ID is '%s'."`

Comment (hannes-ucsc) [inline src/toil/jobStores/abstractJobStore.py]: And space between % and self.

File: src/toil/worker.py [hunk 16]
```diff
@@ -133,9 +134,7 @@ def badWorker():
 
     setLogLevel(config.logLevel)
 
-    tempRootDir = config.workDir
-    if tempRootDir is not None and not os.path.exists(tempRootDir):
-        raise RuntimeError("The temporary directory specified by workDir: %s does not exist" % tempRootDir)
+    tempRootDir = getToilWorkflowDir(config.workflowID, config.workDir)
```
Comment (hannes-ucsc) [inline src/toil/worker.py]: Should we rename tempRootDir or would that be too invasive, i.e. affect many lines in the code?

File: src/toil/jobStores/abstractJobStore.py (hunk 15 above)
Comment (arkal) [inline src/toil/jobStores/abstractJobStore.py]: No i believe this should be INFO such that the user is always told what the workflow ID is. They will need it to access their data in the `toil-<workflowID>` folder

File: src/toil/worker.py (hunk 16 above)
Comment (arkal) [inline src/toil/worker.py]: Just the next statement. Renaming to toilWorkflowDir

File: src/toil/jobStores/abstractJobStore.py (hunk 15 above)
Comment (arkal) [inline src/toil/jobStores/abstractJobStore.py]: Unless you feel the only reason one would want to know the workflow dir is in a debugging situation

**Comment by arkal** [comment]
@hannes-ucsc this is ready for review

**Comment by arkal** [comment]
Also, I checked. The job store is loaded only once when toil is being setup on the leader. Thus the "The workflow ID is: XYZ" message is only printed once. I say we just keep this in INFO

**Comment by hannes-ucsc** [comment]
Then you didn't check thoroughly enough:

https://github.com/arkal/toil/blob/6901c6ec7af964f4d005a6389e3ce672fac1393f/src/toil/worker.py#L102

**Comment by arkal** [comment]
Correction* The job store is loaded only once with config!=None when toil is being setup on the leader. So the workflow ID is created only once.

I moved the logging into the if block so it prints only once.

**Comment by arkal** [comment]
This also means it won't be printed on a restart, though

File: src/toil/jobStores/abstractJobStore.py [hunk 17]
```diff
@@ -77,6 +77,7 @@ def __init__(self, config=None):
         else:
             assert config.workflowID is None
             config.workflowID = str(uuid4())
+            logger.info('The workflow ID is: \'%s\'' % config.workflowID)
```
Comment (hannes-ucsc) [inline src/toil/jobStores/abstractJobStore.py]: `logger.info("The workflow ID is: '%s'" % config.workflowID)`

**Comment by arkal** [comment]
Do you prefer double quotes for this codebase? I default to singles but I can try and stick to doubles from now if you'd like.

**Comment by hannes-ucsc** [comment]
See bullet 3 in https://github.com/BD2KGenomics/toil/wiki/Python-Dos-and-Don'ts

My rule of thumb is to use the least noisy syntax. Backslashes are noisier that double quotes. Double quotes are noisier than single quotes.

**Comment by arkal** [comment]
Got it!

**Pull request merged**